chips/lowrisc/src/flash_ctrl.rs

```diff
@@ -0,0 +1,537 @@
+                    });
+                } else {
+                    // Still writing data, keep trying
+                    self.write_buf.replace(buf);
+                    self.enable_interrupts();
+                }
+            }
+        }
+    }
+}
+
+impl<C: hil::flash::Client<Self>> hil::flash::HasClient<'static, C> for FlashCtrl<'_> {
+    fn set_client(&self, client: &'static C) {
+        self.flash_client.set(client);
+    }
+}
+
+impl hil::flash::Flash for FlashCtrl<'_> {
+    type Page = LowRiscPage;
+
+    fn read_page(
+        &self,
+        page_number: usize,
+        buf: &'static mut Self::Page,
+    ) -> Result<(), (ReturnCode, &'static mut Self::Page)> {
+        let addr = page_number * PAGE_SIZE;
+
+        if !self.data_configured.get() {
+            // If we aren't configured yet, configure now
+            self.configure_data_partition(self.region_num);
+        }
+
+        if !self.info_configured.get() {
+            // If we aren't configured yet, configure now
+            self.configure_info_partition(FlashBank::BANK0, self.region_num);
+        }
+
+        // Enable interrupts and set the FIFO level
+        self.enable_interrupts();
+        self.registers.fifo_lvl.modify(FIFO_LVL::RD.val(0xF));
+
+        // Set the address
+        self.registers.addr.write(ADDR::START.val(addr as u32));
+
+        // Save the buffer
+        self.read_buf.replace(buf);
+        self.read_index.set(0);
+
+        // Start the transaction
+        let info_sel = if self.region_num as u32 > 3 {
+            CONTROL::INFO_SEL::SET
+        } else {
+            CONTROL::INFO_SEL::CLEAR
+        };
+        self.registers.control.write(
+            CONTROL::OP::READ
+                + CONTROL::PARTITION_SEL::DATA
+                + CONTROL::NUM.val(((PAGE_SIZE / 4) - 1) as u32)
+                + CONTROL::START::SET
+                + info_sel,
+        );
+
+        Ok(())
+    }
+
+    fn write_page(
+        &self,
+        page_number: usize,
+        buf: &'static mut Self::Page,
+    ) -> Result<(), (ReturnCode, &'static mut Self::Page)> {
+        let addr = page_number * PAGE_SIZE;
+
+        if !self.data_configured.get() {
+            // If we aren't configured yet, configure now
+            self.configure_data_partition(self.region_num);
+        }
+
+        if !self.info_configured.get() {
+            // If we aren't configured yet, configure now
+            self.configure_info_partition(FlashBank::BANK0, self.region_num);
+        }
+
+        // Erase the page first
+        self.erase_page(page_number);
+
+        // Set the address
+        self.registers.addr.write(ADDR::START.val(addr as u32));
+
+        // Reset the write index
+        self.write_index.set(0);
+
+        // Start the transaction
+        let info_sel = if self.region_num as u32 > 3 {
+            CONTROL::INFO_SEL::SET
+        } else {
+            CONTROL::INFO_SEL::CLEAR
+        };
```

Comment:
I believe this is no longer needed since we are now writing to the data partition?